### PR TITLE
Make it possible to not only write user data but also to modify it

### PR DESCRIPTION
**Pull request by dimitry-**
Bug: http://b/189145024

### CODE REVIEW

File: src/CaptureFile/CaptureFileHelpers.cpp [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+  OUTCOME_TRY(capture_file, CaptureFile::OpenForReadWrite(capture_file_path));
+
+  uint32_t message_size = user_defined_capture_info.ByteSizeLong();
+  const size_t buf_size =
+      message_size + google::protobuf::io::CodedOutputStream::VarintSize32(message_size);
+  auto buf = make_unique_for_overwrite<uint8_t[]>(buf_size);
+  google::protobuf::io::ArrayOutputStream array_output_stream{buf.get(),
+                                                              static_cast<int>(buf_size)};
+  google::protobuf::io::CodedOutputStream coded_output_stream{&array_output_stream};
+  coded_output_stream.WriteVarint32(message_size);
```
Comment (dpallotti) [inline src/CaptureFile/CaptureFileHelpers.cpp:25]: Do we want to document somewhere that we prepend the message with its size? Maybe around here with a comment?

File: src/CaptureFile/ProtoSectionInputStreamImpl.cpp [hunk 2]
```diff
@@ -2,33 +2,42 @@
...
   // Note that in case there was an error CodedInputStream does not provide error messages/codes.
   // We need to go to underlying stream (file_fragment_input_stream_ in this case) to get the error
   // message in case of a failure.
-  if (!coded_input_stream_.ReadVarint64(&message_size)) {
+  if (!coded_input_stream_.ReadVarint32(&message_size)) {
     return file_fragment_input_stream_.GetLastError().value_or(
         ErrorMessage{"Unexpected end of section while reading message size"});
   }
 
+  // Since file input is not trusted having too big value here may lead to out-of-memory allocation
```
[the remark this answers is not in the document]
Comment (dimitry-) [reply src/CaptureFile/ProtoSectionInputStreamImpl.cpp:25]: done.

File: src/CaptureFile/ProtoSectionInputStreamImpl.cpp [hunk 3]
```diff
@@ -2,33 +2,42 @@
...
+  // below. Do a sanity check for message size, we limit our messages to 1Mb maximum size.
+  if (message_size > kMaximumMessageSize) {
+    return ErrorMessage{
+        absl::StrFormat("The message size is too big %d (maximum allowed message size is %d)",
```
[the remark this answers is not in the document]
Comment (dimitry-) [reply src/CaptureFile/ProtoSectionInputStreamImpl.cpp:29]: done.

File: src/CaptureFile/include/CaptureFile/ProtoSectionInputStream.h [hunk 4]
```diff
@@ -2,33 +2,34 @@
...
+#define CAPTURE_FILE_PROTO_SECTION_INPUT_STREAM_H_
+
+#include <google/protobuf/message.h>
 
 #include "OrbitBase/Result.h"
-#include "capture.pb.h"
 
 namespace orbit_capture_file {
 
 // This class is used to read from the main section of capture file.
```
[the remark this answers is not in the document]
Comment (dimitry-) [reply src/CaptureFile/include/CaptureFile/ProtoSectionInputStream.h:14]: yes - fixed.

File: src/CaptureFile/include/CaptureFile/ProtoSectionInputStream.h [hunk 5]
```diff
@@ -2,33 +2,34 @@
...
  public:
-  CaptureSectionInputStream() = default;
-  virtual ~CaptureSectionInputStream() = default;
+  ProtoSectionInputStream() = default;
+  virtual ~ProtoSectionInputStream() = default;
 
-  // Reads next event from the stream. Note that the callser must not read past
-  // orbit_grpc_protos::CaptureFinished message. Doing so will result in
-  // undefined behavior.
+  // Reads next event from the stream to message. Note that the caller must not read past
```
[the remark this answers is not in the document]
Comment (dimitry-) [reply src/CaptureFile/include/CaptureFile/ProtoSectionInputStream.h:21]: yes - fixed.

File: src/CaptureFile/include/CaptureFile/ProtoSectionInputStream.h [hunk 6]
```diff
@@ -2,33 +2,34 @@
...
+  // orbit_grpc_protos::CaptureFinished message in the case of Capture Section. Doing so
+  // will result in undefined behavior.
   //
   // This is because the capture section does not have size and is bounded by the
   // start of the next section and start of all sections are aligned to 8bytes.
   // This means reading after CaptureFinished message sometimes end up reading
   // padded zeros which yield an empty message, or it could generate end of section
   // error.
-  virtual ErrorMessageOr<orbit_grpc_protos::ClientCaptureEvent> ReadEvent() = 0;
+  virtual ErrorMessageOr<void> ReadEvent(google::protobuf::Message* message) = 0;
```
[the remark this answers is not in the document]
Comment (dimitry-) [reply src/CaptureFile/include/CaptureFile/ProtoSectionInputStream.h:30]: done.

File: src/CaptureFile/CaptureFile.cpp [hunk 7]
```diff
@@ -299,6 +298,57 @@ std::optional<uint64_t> CaptureFileImpl::FindSectionByType(uint64_t section_type
...
+    return ErrorMessage{absl::StrFormat(
+        "Cannot resize section %d: The section is located before section list.", section_number)};
+  }
+
+  if (std::any_of(section_list_.begin(), section_list_.end(),
+                  [&section](const CaptureFileSection& section_to_check) {
+                    return section_to_check.offset > section.offset;
+                  })) {
+    return ErrorMessage{absl::StrFormat(
+        "Cannot resize section %d: The section is the last section in the file.", section_number)};
```
[the remark this answers is not in the document]
Comment (dimitry-) [reply src/CaptureFile/CaptureFile.cpp:324]: done.

File: src/CaptureFile/CaptureFileHelpers.cpp (hunk 1 above)
Comment (dimitry-) [reply src/CaptureFile/CaptureFileHelpers.cpp:25]: Added a section to FORMAT.md file.

File: src/CaptureFile/CaptureFileHelpers.cpp [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+  // We do not expect any errors from coded output stream backed by ArrayOutputStream of correct
```
[the remark this answers is not in the document]
Comment (dimitry-) [reply src/CaptureFile/CaptureFileHelpers.cpp:26]: done.

File: src/CaptureFile/include/CaptureFile/CaptureFile.h [hunk 9]
```diff
@@ -33,12 +33,20 @@ class CaptureFile {
...
 
-  // Adds a section, returns added section number. Sections are always added to the end of the list.
-  // The file layout is adjusted accordingly. In case of an I/O error the format consistency is
-  // preserved, but the file size could still end up being changed (for example if updated section
-  // list was successfully written to file and space for the section was successfully reserved
-  // but the function has failed to update file header with the new position of the section list).
-  virtual ErrorMessageOr<uint64_t> AddSection(uint64_t section_type, uint64_t section_size) = 0;
+  // Adds user data section, returns added section number. The user data section added to the end of
+  // the section list. The file layout is adjusted accordingly. This function makes the best effort
+  // to preserve the format consistency in the case of an I/O error  but the file size could still
```
[the remark this answers is not in the document]
Comment (dimitry-) [reply src/CaptureFile/include/CaptureFile/CaptureFile.h:38]: done.

File: src/CaptureFile/include/CaptureFile/CaptureFile.h [hunk 10]
```diff
@@ -33,12 +33,20 @@ class CaptureFile {
   // Returns the index for the first section with specified type.
   [[nodiscard]] virtual std::optional<uint64_t> FindSectionByType(uint64_t section_type) const = 0;
 
-  // Adds a section, returns added section number. Sections are always added to the end of the list.
-  // The file layout is adjusted accordingly. In case of an I/O error the format consistency is
-  // preserved, but the file size could still end up being changed (for example if updated section
-  // list was successfully written to file and space for the section was successfully reserved
-  // but the function has failed to update file header with the new position of the section list).
-  virtual ErrorMessageOr<uint64_t> AddSection(uint64_t section_type, uint64_t section_size) = 0;
+  // Adds user data section, returns added section number. The user data section added to the end of
```
[the remark this answers is not in the document]
Comment (dimitry-) [reply src/CaptureFile/include/CaptureFile/CaptureFile.h:36]: done.

File: src/CaptureFile/include/CaptureFile/CaptureFile.h [hunk 11]
```diff
@@ -33,12 +33,20 @@ class CaptureFile {
...
+  // the section list. The file layout is adjusted accordingly. This function makes the best effort
+  // to preserve the format consistency in the case of an I/O error  but the file size could still
+  // end up being changed (for example if updated section list was successfully written to file and
+  // space for the section was successfully reserved but the function has failed to update file
+  // header with the new position of the section list).
+  virtual ErrorMessageOr<uint64_t> AddUserDataSection(uint64_t section_size) = 0;
+
+  // Extend the last section in the file. This function is intended as fast-path for USER_DATA
+  // read-write section, other sections in the file are supposed to read-only which lets us
+  // to avoid copying data around for the most of the file in the case when only user data
```
[the remark this answers is not in the document]
Comment (dimitry-) [reply src/CaptureFile/include/CaptureFile/CaptureFile.h:46]: done.

File: src/CaptureFile/ProtoSectionInputStreamImpl.cpp [hunk 12]
```diff
@@ -2,33 +2,42 @@
...
+  if (!coded_input_stream_.ReadVarint32(&message_size)) {
     return file_fragment_input_stream_.GetLastError().value_or(
         ErrorMessage{"Unexpected end of section while reading message size"});
   }
 
+  // Since file input is not trusted, having too big value here may lead to out-of-memory allocation
+  // below. Do a sanity check for message size, we limit our messages to 1Mb maximum size.
+  if (message_size > kMaximumMessageSize) {
+    return ErrorMessage{
+        absl::StrFormat("The message size  %d is too big (maximum allowed message size is %d)",
```
Comment (dpallotti) [inline src/CaptureFile/ProtoSectionInputStreamImpl.cpp:29]:
```suggestion
        absl::StrFormat("The message size %d is too big (maximum allowed message size is %d)",
```

**Review by dpallotti — APPROVED** [review]

File: src/CaptureFile/ProtoSectionInputStreamImpl.cpp (hunk 12 above)
Comment (dimitry-) [reply src/CaptureFile/ProtoSectionInputStreamImpl.cpp:29]: done.

**Pull request merged**